pkg/globalnet/controllers/endpoints_controller.go

```diff
@@ -0,0 +1,124 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/federate"
+	"github.com/submariner-io/admiral/pkg/syncer"
+	"github.com/submariner-io/admiral/pkg/util"
+	"github.com/submariner-io/submariner/pkg/globalnet/constants"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+func NewEndpointsController(config *syncer.ResourceSyncerConfig) (Interface, error) {
+	// We'll panic if config is nil, this is intentional
+	var err error
+
+	klog.Info("Creating Endpoints controller")
+
+	_, gvr, err := util.ToUnstructuredResource(&corev1.Endpoints{}, config.RestMapper)
+	if err != nil {
+		return nil, errors.Wrap(err, "error converting resource")
+	}
+
+	controller := &endpointsController{
+		baseSyncerController: newBaseSyncerController(),
+		endpoints:            config.SourceClient.Resource(*gvr),
+	}
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:            "Endpoints syncer",
+		ResourceType:    &corev1.Endpoints{},
+		SourceClient:    config.SourceClient,
+		SourceNamespace: corev1.NamespaceAll,
+		RestMapper:      config.RestMapper,
+		Federator:       federate.NewCreateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Transform:       controller.process,
+	})
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating the syncer")
+	}
+
+	return controller, nil
+}
+
+func (c *endpointsController) Stop() {
+	c.baseController.Stop()
+}
+
+func (c *endpointsController) Start() error {
+	err := c.baseSyncerController.Start()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *endpointsController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	ep := from.(*corev1.Endpoints)
+
+	// Skip endpoints without "endpoints.submariner.io/exported: true" label
```

Comment:
This logic works with an assumption that user will create original endpoints with a specific label `endpoints.submariner.io/exported`. But its very likely that user might miss adding this label to the original endpoint as they will just create regular service without selectors and an associated endpoint. IMHO we should make that use-case work out of the box without expecting this label from the user.

Reply:
We need some means to know if endpoint is created by user for Exported Service or not. If no label, then we have to check if ServiceExport is created or not. Also, we have to handle case where ServiceExport is created after Endpoint is created.

Requirement to have label, though minor, is not without precedent in kubernetes. We can handle Update where one may add label after creation when they decide to export service.

Still, we can't add controller for *all* endpoints as there will be way too many endpoints in a cluster. Note that kube-system endpoint controller adds a label to all endpoints created by it, as does lighthouse for endpointslices created by it. We can easily add code in `subctl` export service to add label for external services.

Reply:
> Still, we can't add controller for _all_ endpoints as there will be way too many endpoints in a cluster. Note lighthouse ads label for endpointslices created by it. We can easily add code in `subctl` export service to add label for external services.

We don't need a controller for all endpoints. We can create a controller/listener with a label selector (just like how we are creating the pod listeners) where the selector will match the name of the original exported service.

Reply:
@sridhargaddam 

I've changed the codes to handle only Endpoints for exported Services. Current results of unit tests shows that it works well regardless of the order of exporting Services and creating Endpoints, however, it doesn't seems to work if original Endpoints resource is updated after cloned Endpoints are created by the controller (in this case, the cloned Endpoints is also expected to be updated to have the same Subsets, but it doesn't seems to be updated).

Could you check if I'm doing something wrong?
(Are there any requirements for admiral to actually update the resource? I assume that just returning the expected resource in `Transform` function makes the resource updated).

Reply:
@tpantelis @sridhargaddam 

I noticed there is `NewCreateOrUpdateFederator`, not `NewCreateFederator`. Is my understanding right that we need to use `NewCreateOrUpdateFederator` if updating resource is required?
If so, what needs to be passed as `localClusterID` and `keepMetadataField`?
(At a glance, just passing "" and nothing seems working, but I need to ensure that it is the right way).

Reply:
> @tpantelis @sridhargaddam
> 
> I noticed there is `NewCreateOrUpdateFederator`, not `NewCreateFederator`. Is my understanding right that we need to use `NewCreateOrUpdateFederator` if updating resource is required? 

Yes

If so, what needs to be passed as `localClusterID` and `keepMetadataField`? (At a glance, just passing "" and nothing seems working, but I need to ensure that it is the right way).

`localClusterID` is used for 2-way syncing with the broker - that's not relevant here so pass "". You also don't need any `keepMetadataField`.

Reply:
@tpantelis Thank you for your quick answer. I will update the code as so.

Reply:
@sridhargaddam  I've updated the PR and it now can handle update. PTAL
 
The last commit is needed for deleting cloned Endpoints if original Endpoints is deleted while controller stops, but it avoids controller become running unless RBAC isn't granted (as handled in submariner-io/submariner-operator#1845 , so it needs to be merged, first).